Specutils 2 compatibility

rosteen opened the pull request:
Replaces #2762 after a significant rebase, with a better title and branch name that are more accurate to the scope of the PR. Still in draft, need to update tests

kecnry commented on jdaviz/configs/cubeviz/plugins/mixins.py:
```diff
@@ -63,7 +63,7 @@ def _set_slice_indicator_value(self, value):
 
 class WithSliceSelection:
-    @cached_property
+    @property
```
is there a reason we need to lose the caching?

rosteen replied:
It seemed to be caching the wrong axis and this worked as a quick fix, I'd like to look into it a bit more once I resolve a few more other things.

rosteen commented:
@cshanahan1 @gibsongreen  this is ready for testing, but will need upstream releases before merging so I'm going to leave it as draft for now. You will need the following to successfully run Jdaviz with this branch:

- The [specutils 2.0 dev branch](https://github.com/astropy/specutils/tree/v2.0-dev)
- The specutils 2 [compatibility PR](https://github.com/glue-viz/glue-astronomy/pull/100) in glue-jupyter 
- Dev install of the `main` branch of specreduce

gibsongreen commented on jdaviz/app.py:
```diff
@@ -771,17 +774,11 @@ def _link_new_data(self, reference_data=None, data_to_be_linked=None):
-            ref_flux_component = dc[0].components[6]
-            linked_wavelength_component = dc[-1].components[1]
-            linked_flux_component = dc[-1].components[-1]
+            ref_wavelength_component = dc[0].components[spectral_axis_index]
+            # May need to update this for specutils 2
+            linked_wavelength_component = linked_data.components[1]
```
Is this why the 2 links go away in the test below:
jdaviz/configs/cubeviz/plugins/moment_maps/tests/test_moment_maps.py

rosteen replied:
I believe so.

gibsongreen commented on jdaviz/app.py:
```diff
@@ -771,17 +774,11 @@ def _link_new_data(self, reference_data=None, data_to_be_linked=None):
-            ]
-
-            dc.add_link(links)
+            dc.add_link(LinkSame(ref_wavelength_component, linked_wavelength_component))
```
Any consequence to going from LinkSameWithUnits to LinkSame?

gibsongreen commented on jdaviz/configs/cubeviz/plugins/slice/slice.py:
```diff
@@ -143,7 +143,8 @@ def _initialize_location(self, *args):
-            if str(viewer.state.x_att) not in self.valid_slice_att_names:
+            if (str(viewer.state.x_att) not in self.valid_slice_att_names and
+                    str(viewer.state.x_att_pixel) not in self.valid_slice_att_names):
                 # avoid setting value to degs, before x_att is changed to wavelength, for example
```
Should this comment be updated?

rosteen replied:
I don't think so.

gibsongreen commented on jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py:
```diff
@@ -495,7 +500,7 @@ def _extract_from_aperture(self, cube, uncert_cube, mask_cube, aperture,
             wcs = cube.coords.spectral_wcs
         else:
-            wcs = None
+            wcs = cube.coords
```
Could this be in an elif? Or cube.coords can be None and will default to that anyways?

gibsongreen commented on jdaviz/configs/cubeviz/plugins/tests/test_parsers.py:
```diff
@@ -197,23 +196,24 @@ def test_numpy_cube(cubeviz_helper):
-    custom_spec = Spectrum1D(flux=[[[20, 1], [9, 1]], [[3, 1], [6, np.nan]]] * u.Unit("erg / Angstrom s cm**2 spaxel"),  # noqa
-                             spectral_axis=[1, 2]*u.AA,
-                             mask=[[[1, 0], [0, 0]], [[0, 0], [0, 0]]])
+    custom_spec = Spectrum(flux=[[[20, 1], [9, 1]], [[3, 1], [6, np.nan]]] * u.Unit("erg / Angstrom s cm**2 spaxel"),  # noqa
```
This wasn't introduced in this PR but is the unit convention spaxel or pix^2?

rosteen replied:
I think that "spaxel" is "spectral pixel" which has a slightly different connotation than just "pix^2", which could just be a square pixel in an image rather than in a slice of a spectral cube.

gibsongreen commented on jdaviz/configs/cubeviz/plugins/tools.py:
```diff
@@ -147,22 +147,30 @@ def _mouse_move_worker(self, x, y):
 
-        if x >= spectrum.flux.shape[0] or x < 0 or y >= spectrum.flux.shape[1] or y < 0:
+        if spectrum.spectral_axis_index in [2, -1]:
+            x_ind = 0
+        else:
+            x_ind = 2
```
is there any case where x_ind can be 1?

rosteen replied:
It's theoretically possible, but (I think) much much less likely than being first or last. I should *probably* code in that possibility, I may try to see if I can centralize this logic somehow so it's not repeated everywhere.

gibsongreen commented on jdaviz/configs/default/plugins/model_fitting/fitting_backend.py:
```diff
@@ -236,13 +240,15 @@ class SpaxelWorker:
     """
-    def __init__(self, flux_cube, wave_array, initial_model, param_set, window=None, mask=None):
+    def __init__(self, flux_cube, wave_array, initial_model, param_set, window=None, mask=None,
+                 spectral_axis_index=2):
```
Any consequence defaulting to 2 if it's not set below? I know this is done in a handful of cases, I'm not entirely familiar with when it should be None versus 2 (below are the examples of it defaulting to both values) jdaviz/configs/default/plugins/model_fitting/initializers.py

rosteen replied:
No, this only ever gets called internally and when it's always set (except that I missed a line where it should be set in one case, fixing that now).

gibsongreen commented:
I'm going to continue going through and testing this out, I did see this traceback in Specviz2d on data load, I'll try out a few more spectra but so far haven't been able to load 2d spectra yet (could be a package issue or something unrelated):

```
TypeError                                 Traceback (most recent call last)
File [~/anaconda3/envs/Work2/lib/python3.12/site-packages/ipyvue/VueTemplateWidget.py:60](http://localhost:8889/~/anaconda3/envs/Work2/lib/python3.12/site-packages/ipyvue/VueTemplateWidget.py#line=59), in Events._handle_event(self, _, content, buffers)
     58     getattr(self, "vue_" + event)(data, buffers)
     59 else:
---> 60     getattr(self, "vue_" + event)(data)

File [~/Work2/jdaviz/jdaviz/core/template_mixin.py:457](http://localhost:8889/~/Work2/jdaviz/jdaviz/core/template_mixin.py#line=456), in with_spinner.<locals>.decorator.<locals>.wrapper(self, *args, **kwargs)
    455 setattr(self, spinner_traitlet, True)
    456 try:
--> 457     ret_ = meth(self, *args, **kwargs)
    458 except Exception:
    459     setattr(self, spinner_traitlet, False)

File [~/Work2/jdaviz/jdaviz/core/loaders/resolvers/resolver.py:270](http://localhost:8889/~/Work2/jdaviz/jdaviz/core/loaders/resolvers/resolver.py#line=269), in BaseResolver.vue_import_clicked(self, *args, **kwargs)
    268 @with_spinner('import_spinner')
    269 def vue_import_clicked(self, *args, **kwargs):
--> 270     self.importer()

File [~/Work2/jdaviz/jdaviz/core/loaders/importers/spectrum2d/spectrum2d.py:124](http://localhost:8889/~/Work2/jdaviz/jdaviz/core/loaders/importers/spectrum2d/spectrum2d.py#line=123), in Spectrum2DImporter.__call__(self)
    121 data_label = self.data_label_value
    122 ext_data_label = self.ext_data_label_value
--> 124 super().__call__()
    126 if not self.auto_extract:
    127     return

File [~/Work2/jdaviz/jdaviz/core/loaders/importers/importer.py:154](http://localhost:8889/~/Work2/jdaviz/jdaviz/core/loaders/importers/importer.py#line=153), in BaseImporterToDataCollection.__call__(self)
    152 if self.data_label_invalid_msg:
    153     raise ValueError(self.data_label_invalid_msg)
--> 154 self.add_to_data_collection(self.output, show_in_viewer=True)

File [~/Work2/jdaviz/jdaviz/core/loaders/importers/spectrum2d/spectrum2d.py:117](http://localhost:8889/~/Work2/jdaviz/jdaviz/core/loaders/importers/spectrum2d/spectrum2d.py#line=116), in Spectrum2DImporter.output(self)
    114 else:
    115     kw = {'wcs': wcs}
--> 117 return Spectrum(flux=data * data_unit, meta=metadata, **kw)

TypeError: unsupported operand type(s) for *: 'NoneType' and 'IrreducibleUnit'
```

rosteen commented:
> I'm going to continue going through and testing this out, I did see this traceback in Specviz2d on data load, I'll try out a few more spectra but so far haven't been able to load 2d spectra yet (could be a package issue or something unrelated):

Hm, the Specviz2D example notebook loads the data fine for me. Do you have an up to date `main` branch of specreduce installed?

gibsongreen commented:
> Hm, the Specviz2D example notebook loads the data fine for me. Do you have an up to date `main` branch of specreduce installed?

Just upgraded and wasn't that but I'll continue playing around with the packages I hadn't touched this env in a few months so it's more then likely that

cshanahan1 reviewed (commented):
there are still references to Spectrum1D in import_data.rst, and cubeviz_aperture_photometry.ipynb

cshanahan1 reviewed (commented):
Testing this branch on the Specviz example notebook I ran into:

'TypeError: Could not find a class to translate objects of type Spectrum to Data'.

rosteen commented:
> Testing this branch on the Specviz example notebook I ran into:
> 
> 'TypeError: Could not find a class to translate objects of type Spectrum to Data'.

It sounds like you don't have the glue-astronomy PR that's needed, see the [comment above](https://github.com/spacetelescope/jdaviz/pull/2922#issuecomment-2807073921)

cshanahan1 commented on jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:
```diff
@@ -606,6 +614,12 @@ def calculate_photometry(self, dataset=None, aperture=None, background=None,
+        if self.config == "cubeviz" and data.ndim > 2:
+            if "spectral_axis_index" in getattr(data, "meta", {}):
+                spectral_axis_index = data.meta["spectral_axis_index"]
+            else:
```
when would case occur, where "spectral_axis_index" is not in meta and it defaults in this plugin to 0?

cshanahan1 commented on jdaviz/configs/imviz/plugins/coords_info/coords_info.py:
```diff
@@ -65,6 +65,7 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self._marks = {}
         self._dict = {}  # dictionary representation of current mouseover info
+        self._spectral_axis_index = 2  # Needed for cube data
```
any consequences of this defaulting to 2 initially?

cshanahan1 reviewed (commented):
left a few suggestions / requests for clarification. some changes i don't fully understand the implications of (for example, why .icrs is not needed when spectral axis=0, and some others i left comments asking for clarification) but I will trust you on the nitty gritty of those. I messed around with specviz and specviz2d and the basic workflows I tried all worked so far.

There are a lot of test failures when running the tests locally with the correct specutils/specreduce/glue-astronomy branches, so I will re-review once this is out of draft and those are passing. Can you pin those versions in this branch (i think you can point to your glue astronomy branch)?

rosteen commented on jdaviz/configs/default/plugins/model_fitting/model_fitting.py:
```diff
@@ -545,42 +546,57 @@ def _initialize_model_component(self, model_comp, comp_label, poly_order=None):
                 data = self.app.data_collection[0].get_object(statistic=None)
-            masked_spectrum = self._apply_subset_masks(data, self.spectral_subset)
+            spatial_axes = [0, 1, 2]
+            spatial_axes.remove(data.spectral_axis_index)
```
If the spectral_axis_index is input as -1, specutils internally changes it to be 2 (in this case, whatever is appropriate). So I think it's ok.

rosteen commented:
FYI in the [last commit](https://github.com/spacetelescope/jdaviz/pull/2922/commits/ffeef5e7e0875e3818e8abfc0bc361126823adc1) I fixed (I'm pretty sure it was wrong and is now corrected anyway) a test fixture where the WCS was corresponding to the wrong axes in a 2D spectrum. Previously it was being made wrong and then Mosviz was transposing it to put the long axis on the x-axis, so the flux values had to be updated since they're now natively getting generated with the right shape and not transposed.

kecnry commented on jdaviz/app.py:
```diff
@@ -790,6 +791,8 @@ def _link_new_data(self, reference_data=None, data_to_be_linked=None):
         if self.config == 'mosviz':
             # In Mosviz, first data is always MOS Table. Use the next data
             default_refdata_index = 1
+        elif self.config == 'cubeviz':
```
I would love for this to do something other than check the config (thinking of deconfigging efforts), but everything else in this block follows the same pattern, so probably makes sense to defer that

kecnry commented on jdaviz/app.py:
```diff
@@ -1121,15 +1126,15 @@ def get_subsets(self, subset_name=None, spectral_only=False,
     def _is_subset_spectral(self, subset_region):
         if isinstance(subset_region, SpectralRegion):
             return True
-        elif isinstance(subset_region, list) and len(subset_region) > 0:
+        elif isinstance(subset_region, list):
```
why the change here?  Could there still be a list with no length and what would happen now?

kecnry commented on jdaviz/configs/cubeviz/plugins/cube_listener.py:
```diff
@@ -63,14 +67,24 @@ def sonify_spectrum(spec, duration, overlap=0.05, system='mono', srate=44100, fm
         self.maxval = pow(2, bdepth-1) - 1
         self.fadedx = 0
+        if n_cpu is None:
+            self.n_cpu = mp.cpu_count() - 1
```
is it ever possible that this would result in `n_cpu = 0` and we should worry about that?  Is subtracting the one to leave one CPU for the manager/kernel?

kecnry commented on jdaviz/configs/cubeviz/plugins/cube_listener.py:
```diff
@@ -119,29 +135,50 @@ def sonify_cube(self):
         self.newsig[:] = self.cursig[:]
         t1 = time.time()
-        print(f"Took {t1-t0}s to process {self.cube.shape[0]*self.cube.shape[1]} spaxels")
+        print(f"Took {t1-t0}s to process "
+              f"{self.cube.shape[self.spatial_inds[0]]*self.cube.shape[self.spatial_inds[1]]} spaxels")  # noqa
```
can we make this a snackbar while we're at it?

kecnry commented on jdaviz/configs/cubeviz/plugins/moment_maps/moment_maps.py:
```diff
@@ -305,15 +303,17 @@ def calculate_moment(self, add_data=True):
 
         # Finally actually calculate the moment
-        self.moment = analysis.moment(slab, order=n_moment).T
+        self.moment = analysis.moment(slab, order=n_moment)
```
is this because `Spectrum` is being passed in transposed already so the output doesn't need to be?

rosteen replied:
I think it's the opposite, it's because we're _not_ transposing the spectrum anymore. A second visual sanity check (I know I did when I was coding this) to make sure the output moment lines up would be prudent though if you didn't do so.

kecnry commented on jdaviz/configs/cubeviz/plugins/moment_maps/moment_maps.py:
```diff
@@ -357,6 +357,15 @@ def calculate_moment(self, add_data=True):
+        if hasattr(w, 'celestial'):
+            # This is the FITS WCS case
+            data_wcs = getattr(w, 'celestial', None)
+        elif hasattr(w, 'to_fits_sip'):
+            # If it's a GWCS we pull out the celestial part
+            data_wcs = WCS(w.to_fits_sip())
```
do we always want to do this?  For loading in actual data, we only convert to sip if requested by the user.  What change in specutils requires this?

rosteen replied:
This is for the case where the coords are a GWCS, because GWCS doesn't have separable `celestial` and `spectral` attributes that you can just pull out individually (unless I've missed something...). This is the only way I found to get out just the 2D spatial solution from the 3D GWCS.

rosteen replied:
Basically you _can_ make a SIPS approximation of the 3 dimensional cube, but you don't have to because you can just use the 3D GWCS. But this is easiest (only?) way to get the purely 2D spatial solution out of the GWCS that I know of.

kecnry commented on jdaviz/configs/cubeviz/plugins/moment_maps/tests/test_moment_maps.py:
```diff
@@ -155,14 +155,13 @@ def test_moment_calculation(cubeviz_helper, spectrum1d_cube,
+    assert len(dc.external_links) == 5  # pixel linked
     # Link 3D z to 2D x and 3D y to 2D y
-    assert (dc.external_links[2].cids1[0].label == "Pixel Axis 0 [z]" and
-            dc.external_links[2].cids2[0].label == "Pixel Axis 1 [x]" and
-            dc.external_links[3].cids1[0].label == "Pixel Axis 1 [y]" and
-            dc.external_links[3].cids2[0].label == "Pixel Axis 0 [y]")
+    assert (dc.external_links[3].cids1[0].label == "Pixel Axis 1 [y]" and
+            dc.external_links[3].cids2[0].label == "Pixel Axis 0 [y]" and
+            dc.external_links[4].cids1[0].label == "Pixel Axis 2 [x]" and
+            dc.external_links[4].cids2[0].label == "Pixel Axis 1 [x]")
```
this hurts my head, I'm going to trust you

kecnry reviewed (commented):
geez, on second thought maybe we should just stick with `Spectrum1D` (kidding!)

to be continued (still need to finish going through the remainder of the diff from `parsers.py` on but wanted to post some comments to start iterating and avoid risk of them getting lost in draft)

javerbukh commented:
Ran all the configs that use `Spectrum` and could not find any issues, nice work!! @kecnry did have one comment about `len(subset_region)` being removed and I wanted clarification on that. I will also assume all the linking changes make sense since I did not run into any bugs... 🤞

kecnry commented on jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:
```diff
@@ -528,9 +528,17 @@ def _calc_background_median(self, reg, data=None):
 
         comp = data.get_component(data.main_components[0])
+        if self.config == "cubeviz" and data.ndim > 2:
+            if "spectral_axis_index" in getattr(data, "meta", {}):
+                spectral_axis_index = data.meta["spectral_axis_index"]
+            else:
+                spectral_axis_index = 0
```
Too convoluted as a single line? 🤷‍♂️ 

```suggestion
            spectral_axis_index = getattr(data, "meta", {}).get("spectral_axis_index", 0)
```

kecnry commented on jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:
```diff
@@ -528,9 +528,17 @@ def _calc_background_median(self, reg, data=None):
                 data = self.dataset.selected_dc_item
 
         comp = data.get_component(data.main_components[0])
+        if self.config == "cubeviz" and data.ndim > 2:
```
Is it safe to generalize this?  Or do we worry about a case where the extra axis isn't spectral (maybe we can check directly for that instead)?

```suggestion
        if data.ndim > 2:
```

kecnry commented on jdaviz/configs/imviz/plugins/aper_phot_simple/aper_phot_simple.py:
```diff
@@ -632,6 +640,12 @@ def calculate_photometry(self, dataset=None, aperture=None, background=None,
             # we can use the pre-cached value
             data = self.dataset.selected_dc_item
 
+        if self.config == "cubeviz" and data.ndim > 2:
```
(see comment above)

```suggestion
        if data.ndim > 2:
```

kecnry commented on jdaviz/configs/imviz/plugins/coords_info/coords_info.py:
```diff
@@ -65,6 +65,7 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self._marks = {}
         self._dict = {}  # dictionary representation of current mouseover info
+        self._spectral_axis_index = 2  # Needed for cube data
```
will this choke if multiple cubes use different values for `spectral_axis_index`?

rosteen replied:
Actually, this gets set based on the reference data cube when it's actually used, it should be fine.

kecnry commented on jdaviz/configs/imviz/plugins/coords_info/coords_info.py:
```diff
@@ -384,10 +391,17 @@ def _image_viewer_update(self, viewer, x, y, mouseevent=True):
             else:
                 data_wcs = None
 
+            if "spectral_axis_index" in getattr(coo_data, "meta", {}):
+                self._spectral_axis_index = coo_data.meta["spectral_axis_index"]
```
I see this is grabbing from the data... do we need to set it as a class attribute at all then?

rosteen replied:
It does get used in one or two other places, it seemed more convenient to store it rather than access the data meta every time.

kecnry commented on jdaviz/configs/mosviz/tests/test_data_loading.py:
```diff
@@ -221,12 +221,12 @@ def test_load_single_image_multi_spec(mosviz_helper, mos_image, spectrum1d, mos_
-                                        {'event': 'mousemove', 'domain': {'x': 10, 'y': 100}})
+                                        {'event': 'mousemove', 'domain': {'x': 100, 'y': 10}})
 
     # Note: spectra2d Wave loaded in meters, but we respect one spectral unit, so the meters in
     # converted to Angstrom (the spectra1d spectral unit).
-    assert label_mouseover.as_text() == ('Pixel x=00010.0 y=00100.0 Value +8.12986e-01 Jy',
-                                         'Wave 1.10000e+05 Angstrom', '')
+    assert label_mouseover.as_text() == ('Pixel x=00100.0 y=00010.0 Value +1.26544e-01 Jy',
+                                         'Wave 1.01000e+06 Angstrom', '')
```
is this transpose expected or fixing a previous bug?

rosteen replied:
Fixing a previous bug, the data shape was backwards compared to what it should have been in the test fixture given the WCS definition.

rosteen replied:
That's the reason for the changes in  `jdaviz/configs/mosviz/tests/test_tools.py` as well.

kecnry commented on jdaviz/utils.py:
```diff
@@ -480,6 +483,17 @@ def get_subset_type(subset):
                 # subset type:
                 continue
 
+            # check for spectral coordinate in GWCS by looking for SpectralFrame
+            if isinstance(ss_data.coords, gwcs):
+                if isinstance(ss_data.coords, (SpectralFrame, SpectralGWCS)):
+                    return 'spectral'
+                elif isinstance(ss_data.coords, CompositeFrame):
+                    if np.any([isinstance(frame, SpectralFrame) for frame in
+                               ss_data.coords.output_frame.frames]):
+                        return 'spectral'
+                else:
+                    continue
```
doesn't seem possible to get here because of the previous `else: continue`

rosteen replied:
Perhaps but eh, better safe than sorry. I don't remember if I had a specific case that escaped the other handling.

rosteen replied:
It's been long enough that it makes me nervous to remove it 😆

kecnry replied:
but its literally impossible for the code to reach here, right?  If we need to handle this case, should it be somewhere else?

rosteen commented on jdaviz/configs/cubeviz/plugins/tests/test_cubeviz_helper.py:
```diff
@@ -54,5 +54,5 @@ def test_get_data_spatial_and_spectral(cubeviz_helper, spectrum1d_cube_larger):
                                             True, True, True, True, True]
-    assert max(list(spatial_with_spec.flux.value)) == 232.
-    assert min(list(spatial_with_spec.flux.value)) == 16.
+    assert max(list(spatial_with_spec.flux.value)) == 157.
+    assert min(list(spatial_with_spec.flux.value)) == 13.
```
I believe I looked into this and decided the old values were actually incorrect, but I can't pull a detailed explanation out of my head right this moment for what the bug was. @cshanahan1 same answer to your comment above.

rosteen commented on jdaviz/configs/default/plugins/model_fitting/fitting_backend.py:
```diff
@@ -320,30 +333,47 @@ def _build_model(component_list, expression):
     """
-    n_x, n_y, _ = spectrum.flux.shape
+    # Handle this being either a Spectrum or an array passed from sonification
+    if spectral_axis_index is None and hasattr(spectrum, 'spectral_axis_index'):
+        spectral_axis_index = spectrum.spectral_axis_index
```
The intent is that if you're not passing a `Spectrum` you have to specify this. It's only intended to be used internally, but I needed it to be importable from elsewhere so I removed the leading underscore.

javerbukh commented:
I tested it again and it looks like the slow down happens only if you use a spectral subset and sonify just that range. With no subset selected it took 56 seconds (slight speed up) and with a small range selected it took 86 seconds.

rosteen commented on jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py:
```diff
@@ -49,27 +49,27 @@ def test_linking_after_spectral_smooth(cubeviz_helper, spectrum1d_cube):
     assert len(dc) == 3
     assert dc[-1].label == f'{data_label}[FLUX] spectral-smooth stddev-3.2'
-    assert len(dc.external_links) == 5
+    assert len(dc.external_links) == 4
```
I think this was due to replacing a LinkSameWithUnits + LinkSame with just a LinkSame (Gilbert commented on that elsewhere).

rosteen commented on jdaviz/configs/default/plugins/viewers.py:
```diff
@@ -783,24 +783,26 @@ def _plot_uncertainties(self):
-                # This may need adjustment after the following
-                # specutils PR is merged: https://github.com/astropy/specutils/pull/1033
-                spectral_axis = -1
+                spectral_axis = lyr.data.meta['spectral_axis_index']
```
Good point, I added a fallback in case this is absent - does LCViz subclass this for its profile viewer? That's the main case right now I would assume this would have failed.
